### PR TITLE
Expand Blazor logging

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
Implementation for https://github.com/aspnet/AspNetCore/issues/11792, except for the JS interop parts which @javiercn is handling.

### CODE REVIEW

File: src/Components/Components/src/Rendering/Renderer.cs [hunk 1]
```diff
@@ -68,6 +71,8 @@ public Renderer(IServiceProvider serviceProvider)
         public Renderer(IServiceProvider serviceProvider, IDispatcher dispatcher) : this(serviceProvider)
         {
             _dispatcher = dispatcher;
+            _logger = (ILogger<Renderer>)serviceProvider.GetService(typeof(ILogger<Renderer>))
```
Comment (pranavkm) [inline src/Components/Components/src/Rendering/Renderer.cs]: Shouldn't this be done in the other ctor? Also how do you feel about passing in an `ILogger` \ `ILoggerFactory` instead to the ctor? This feels a little gross.

File: src/Components/Components/src/Rendering/Renderer.cs [hunk 2]
```diff
@@ -220,6 +226,8 @@ public virtual Task DispatchEventAsync(int eventHandlerId, EventFieldInfo fieldI
                 throw new ArgumentException($"There is no event handler with ID {eventHandlerId}");
```
Comment (pranavkm) [inline src/Components/Components/src/Rendering/Renderer.cs]: Slightly related - should this log and return? The values are untrusted user input and we don't want it to start spamming error logs

File: src/Components/Components/src/Rendering/Renderer.cs [hunk 3]
```diff
@@ -220,6 +226,8 @@ public virtual Task DispatchEventAsync(int eventHandlerId, EventFieldInfo fieldI
...
             }
 
+            Log.HandlingEvent(_logger, eventHandlerId, eventArgs);
```
Comment (pranavkm) [inline src/Components/Components/src/Rendering/Renderer.cs]: 👍

File: src/Components/Components/src/Rendering/Renderer.cs [hunk 4]
```diff
@@ -220,6 +226,8 @@ public virtual Task DispatchEventAsync(int eventHandlerId, EventFieldInfo fieldI
...
+
             if (fieldInfo != null)
             {
                 var latestEquivalentEventHandlerId = FindLatestEventHandlerIdInChain(eventHandlerId);
```
Comment (pranavkm) [inline src/Components/Components/src/Rendering/Renderer.cs]: Is there any value (for us to debug) to also log this?

File: src/Components/Components/src/Routing/Router.cs [hunk 5]
```diff
@@ -55,6 +59,8 @@ public class Router : IComponent, IHandleAfterRender, IDisposable
         /// <inheritdoc />
         public void Configure(RenderHandle renderHandle)
         {
+            _logger = (ILogger<Router>)ServiceProvider.GetService(typeof(ILogger<Router>))
```
Comment (pranavkm) [inline src/Components/Components/src/Routing/Router.cs]: Is the intent to make this optional in WASM? Could we add a comment to that effect? It's very tempting to optimize it and make this use property injection.

File: src/Components/Server/src/Circuits/RemoteRenderer.cs [hunk 6]
```diff
@@ -179,7 +179,7 @@ private async Task WriteBatchBytesAsync(PendingRender pending)
                     return;
                 }
 
-                Log.BeginUpdateDisplayAsync(_logger, _client.ConnectionId);
+                Log.BeginUpdateDisplayAsync(_logger, _client.ConnectionId, pending.BatchId, pending.Data.Length);
```
Comment (pranavkm) [inline src/Components/Server/src/Circuits/RemoteRenderer.cs]: While you're editing this file, could you update the log message on line 218 to use `LoggerMessage`?

**Review by javiercn — APPROVED** [review]
Pretty straightforward change!

I'm not uber familiar with the define pattern for the Log class, but I assume you've followed previous established practices.

File: src/Components/Server/src/Circuits/RemoteRenderer.cs (hunk 6 above)
Comment (pranavkm) [reply src/Components/Server/src/Circuits/RemoteRenderer.cs]: There's also some in `RemoteUriHelper` and `ComponentHub`:

* https://github.com/aspnet/AspNetCore/blob/master/src/Components/Server/src/ComponentHub.cs#L76
* https://github.com/aspnet/AspNetCore/blob/master/src/Components/Server/src/ComponentHub.cs#L134
* https://github.com/aspnet/AspNetCore/blob/master/src/Components/Server/src/ComponentHub.cs#L143
* https://github.com/aspnet/AspNetCore/blob/master/src/Components/Server/src/ComponentHub.cs#L152
* https://github.com/aspnet/AspNetCore/blob/master/src/Components/Server/src/Circuits/RemoteUriHelper.cs#L63
* https://github.com/aspnet/AspNetCore/blob/master/src/Components/Server/src/Circuits/RemoteUriHelper.cs#L82
* https://github.com/aspnet/AspNetCore/blob/master/src/Components/Server/src/Circuits/RemoteUriHelper.cs#L89

File: src/Components/Components/src/Rendering/Renderer.cs (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Components/Components/src/Rendering/Renderer.cs]: > Shouldn't this be done in the other ctor?

Yes, absolutely right! I knew to do that earlier then for some reason started working on something else and forgot about it. Thanks for catching that.

> Also how do you feel about passing in an ILogger \ ILoggerFactory instead to the ctor? This feels a little gross. This feels a little gross.

I did this because we already take a `serviceProvider`, so why complicate things for the consumer? Under what circumstances would someone have reason to supply a different `ILogger` than the one the service provider would offer? I'm not absolutely set against it, but would like to make a choice based on practical reasoning.

In the long term, we may need to consume other services too. If we keep adding them as extra constructor overloads one-by-one, it will not be a nice experience for consumers.

File: src/Components/Components/src/Rendering/Renderer.cs (hunk 2 above)
Comment (SteveSandersonMS) [reply src/Components/Components/src/Rendering/Renderer.cs]: This exception is treated as a regular JS interop error, so it's no different from someone spamming the server with any kind of invalid inputs to a JS interop call. We'll already be logging those as "info", so that's the same effect as you're suggesting, right?

@javiercn please confirm that matches your expectation of how we'll handle JS interop exceptions.

File: src/Components/Components/src/Rendering/Renderer.cs (hunk 1 above)
Comment (javiercn) [reply src/Components/Components/src/Rendering/Renderer.cs]: Who will override the renderer in the first place other than our platforms?

> I did this because we already take a `serviceProvider`, so why complicate things for the consumer

It makes the dependencies visible

> Under what circumstances would someone have reason to supply a different `ILogger` than the one the service provider would offer

Never, but mostly because it is very unlikely to replace the renderer in the stack

Comment (javiercn) [reply src/Components/Components/src/Rendering/Renderer.cs]: > n the long term, we may need to consume other services too. If we keep adding them as extra constructor overloads one-by-one, it will not be a nice experience for consumers.

I don't think there is an scenario, other than the platform for consuming the renderer.

File: src/Components/Components/src/Rendering/Renderer.cs (hunk 4 above)
Comment (SteveSandersonMS) [reply src/Components/Components/src/Rendering/Renderer.cs]: I don't see any definite reason to do that, but we could always add an extra log message in the future if a reason arises.

File: src/Components/Components/src/Rendering/Renderer.cs (hunk 2 above)
Comment (javiercn) [reply src/Components/Components/src/Rendering/Renderer.cs]: I didn't realize the event was coming form there, but yes. If it comes through JS interop it will be sent back to the client. WRT logging it on the server, I will be adding that on a separate PR

File: src/Components/Components/src/Routing/Router.cs (hunk 5 above)
Comment (SteveSandersonMS) [reply src/Components/Components/src/Routing/Router.cs]: Not really WASM (because that does have a logger by default) but further-out edge cases like unit test harnesses. Do you think there's any disadvantage to being tolerant to that?

File: src/Components/Server/src/Circuits/RemoteRenderer.cs (hunk 6 above)
Comment (SteveSandersonMS) [reply src/Components/Server/src/Circuits/RemoteRenderer.cs]: Sure, good plan.

File: src/Components/Components/src/Rendering/Renderer.cs (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Components/Components/src/Rendering/Renderer.cs]: > Who will override the renderer in the first place other than our platforms?
> I don't think there is an scenario, other than the platform for consuming the renderer.

I don't agree that only people who work for Microsoft will create and/or consume renderers. We have examples of renderers for Electron and Flutter, so it would be surprising if nobody ever had the idea of making or hosting another renderer on a different UI technology, or for use in a different way such as a unit testing system.

> It makes the dependencies visible

Not sure about that.

In terms of *possible* dependencies that may be consumed, the renderer possibly depends on every possible service that an `IServiceProvider` can provide, since it's responsible for instantiating other components that themselves require arbitrary services.

In terms of *required* dependencies, `ILogger<Renderer>` isn't required (it's an optional dependency by the logic here).

File: src/Components/Components/src/Rendering/Renderer.cs (hunk 2 above)
Comment (SteveSandersonMS) [reply src/Components/Components/src/Rendering/Renderer.cs]: Cool, that's what I thought. Considering this resolved then, but please speak up @pranavkm if you have any remaining concerns.

File: src/Components/Components/src/Rendering/Renderer.cs (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Components/Components/src/Rendering/Renderer.cs]: Thinking about this more, this particular detail just doesn’t really bother me either way. There are two of you saying you prefer something else, so I’ll switch this to be a constructor param.

File: src/Components/Components/src/Routing/Router.cs (hunk 5 above)
Comment (pranavkm) [reply src/Components/Components/src/Routing/Router.cs]: Nope, just that it stands out amongst other injected types.

**Review by pranavkm — APPROVED** [review]

File: src/Components/Components/src/Routing/Router.cs (hunk 5 above)
Comment (SteveSandersonMS) [reply src/Components/Components/src/Routing/Router.cs]: OK, I'm switching over to inject an `ILoggerFactory` since it's more conventional. The idea of making the logger optional doesn't really achieve anything.

File: src/Components/Components/src/Rendering/Renderer.Log.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Components.Rendering
+{
+    public abstract partial class Renderer
+    {
+        internal static class Log
+        {
+            private static readonly Action<ILogger, int, Type, int, Type, Exception> _initializingChildComponent =
+                LoggerMessage.Define<int, Type, int, Type>(LogLevel.Trace, new EventId(1, "InitializingChildComponent"), "Initializing component {ComponentId} ({ComponentType}) as child of {ParentComponentId} ({ParentComponentId})");
```
Comment (rynowak) [inline src/Components/Components/src/Rendering/Renderer.Log.cs]: These all should be `LogLevel.Debug`. In practice `Trace` means debugging message that contains **sensitive** data like a user's credit card number. None of these have sensitive data.

File: src/Components/Components/src/Routing/Router.cs (hunk 5 above)
Comment (rynowak) [reply src/Components/Components/src/Routing/Router.cs]: You can also inject `ILogger<Router>` directly which saves some annoying.

**Review by rynowak — COMMENTED** [review]
Changes look good so far 👍

File: src/Components/Components/src/Rendering/Renderer.Log.cs (hunk 7 above)
Comment (rynowak) [reply src/Components/Components/src/Rendering/Renderer.Log.cs]: note: we also don't consider things likes URLs to be sensitive, because it would be hard to log anything comprehensible for users without including them.

Comment (SteveSandersonMS) [reply src/Components/Components/src/Rendering/Renderer.Log.cs]: OK. The docs described a difference based on level of detail which sounded pretty subjective, so I took a guess based on how deep in the system they are.

Happy to switch them to `Debug` if this is the convention.

File: src/Components/Server/src/Circuits/RemoteRenderer.cs [hunk 8]
```diff
@@ -279,10 +279,10 @@ static Log()
                     EventIds.UnhandledExceptionRenderingComponent,
                     "Unhandled exception rendering component: {Message}");
 
-                _beginUpdateDisplayAsync = LoggerMessage.Define<string>(
+                _beginUpdateDisplayAsync = LoggerMessage.Define<long, int, string>(
                     LogLevel.Trace,
                     EventIds.BeginUpdateDisplayAsync,
-                    "Begin remote rendering of components on client {ConnectionId}.");
+                    "Sending render batch {batchId} of size {dataLength} bytes to client {ConnectionId}.");
```
Comment (BrennanConroy) [inline src/Components/Server/src/Circuits/RemoteRenderer.cs]: nit: Use PascalCase for structured log keys

Comment (SteveSandersonMS) [reply src/Components/Server/src/Circuits/RemoteRenderer.cs]: Done, thanks.

File: src/Components/Components/src/Rendering/Renderer.Log.cs (hunk 7 above)
Comment (rynowak) [reply src/Components/Components/src/Rendering/Renderer.Log.cs]: I'm sure the docs say something. In practice we use debug unless it's sensitive data, then we use trace.

**Review by rynowak — APPROVED** [review]
awesome!
